### PR TITLE
add userId configuration option

### PR DESCRIPTION
**Pull request by Xavrax**
feat: add `user_id` configuration option that deprecates `uuid` ones

feat: `PUBNUB_DEPRECATED` macro

BREAKING CHANGES: now `user_id` (old `uuid`) is a required property!

### CODE REVIEW

**Comment by budgetpreneur** [comment]
@Xavrax Could you please update this branch with base branch to continue by tests.

File: qt/pubnub.hpp [hunk 1]
```diff
@@ -279,21 +280,48 @@ namespace pubnub {
...
+    	void set_user_id(std::string const& user_id)
+    	{
+            d_pbqt.set_user_id(QString::fromStdString(uuid));
+    	}
+
+	/// Set the user_id with a random-generated UUID
+        /// 
+	/// @deprecated this is provided as a workaround for existing users.
+        /// Please use `set_user_id_with_random_uuid_v4` instead.
+   	PUBNUB_DEPRECATED int set_uuid_v4_random() {
```
Comment (kleewho) [inline qt/pubnub.hpp:305]: Can you explain to me why sometimes we have `/** */` and sometimes `///`? It's nothing major obviously, I'm just curious

File: core/pubnub_fetch_history.h [hunk 2]
```diff
@@ -43,11 +43,11 @@ struct pubnub_fetch_history_options {
      * false.
      */
     bool include_message_type;
-    /** If true to recieve uuid with each history
-     * message. If false, no uuid per message. Defaults to
+    /** If true to recieve user_id with each history
```
Comment (kleewho) [inline core/pubnub_fetch_history.h:46]: There's a typo which you could fix as well. `recieve` -> `receive`

File: qt/pubnub.hpp (hunk 1 above)
Comment (Xavrax) [reply qt/pubnub.hpp:305]: No idea. 

I'm following the rules that most fit into current file. 

I think that I will merge comment styles in all repository but this PR was made at the beginning of my contributions se I wasn't brave enough to do such a change.

**Comment by budgetpreneur** [comment]
@Xavrax , some memory related issue (not related to user_id changes)
C:\Pandu\GitHub\c-core\openssl>pubnub_encrypt_decrypt_static_iv_sample.exe
Message to be encrypted = yay!

bytes before encoding iv + encrypted msg = [171 252 73 170 163 122 169 184 153 49 118 38 137 0 181 23 ]
base64encode tried to use more than 23 bytes to encode 16 bytes
base64 message = x‗o (size = 7929970)

 *****************************************************
=================================================================
==24444==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x006fed5a at pc 0x00e522ba bp 0x006fed3c sp 0x006fed3c
READ of size 1 at 0x006fed5a thread T0
    #0 0xe522b9 in pbbase64_decode C:\Pandu\GitHub\c-core\lib\base64\pbbase64.c:176
    #1 0xe52a31 in pbbase64_decode_alloc C:\Pandu\GitHub\c-core\lib\base64\pbbase64.c:226
    #2 0xe52f34 in pbbase64_decode_alloc_std C:\Pandu\GitHub\c-core\lib\base64\pbbase64.c:257
    #3 0xe52f88 in pbbase64_decode_alloc_std_str C:\Pandu\GitHub\c-core\lib\base64\pbbase64.c:263
    #4 0xe2ef73 in pubnub_decrypt C:\Pandu\GitHub\c-core\core\pubnub_crypto.c:197
    #5 0xdf16df in main C:\Pandu\GitHub\c-core\core\samples\pubnub_encrypt_decrypt_iv_sample.c:67
    #6 0xe8dd63 in _scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    #7 0x76fdfa28  (C:\WINDOWS\System32\KERNEL32.DLL+0x6b81fa28)
    #8 0x77c27a9d  (C:\WINDOWS\SYSTEM32\ntdll.dll+0x4b2e7a9d)
    #9 0x77c27a6d  (C:\WINDOWS\SYSTEM32\ntdll.dll+0x4b2e7a6d)

Address 0x006fed5a is located in stack of thread T0 at offset 2 in frame
    #0 0xe51dcf in pbbase64_decode C:\Pandu\GitHub\c-core\lib\base64\pbbase64.c:141

  This frame has 2 object(s):
    [16, 272) 'decode_tab'
    [32, 36) 'word'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp, SEH and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-underflow C:\Pandu\GitHub\c-core\lib\base64\pbbase64.c:176 in pbbase64_decode
Shadow bytes around the buggy address:
  0x300dfd50: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x300dfd60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x300dfd70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x300dfd80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x300dfd90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x300dfda0: 00 00 00 00 00 00 00 00 00 00 00[f1]f1 00 00 00
  0x300dfdb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x300dfdc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 f2 f2 f2
  0x300dfdd0: f2 04 f3 f3 f3 f3 00 00 00 00 00 00 00 00 00 00
  0x300dfde0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x300dfdf0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==24444==ABORTING

File: cpp/pubnub_common.hpp [hunk 3]
```diff
@@ -632,30 +633,67 @@ class context {
...
     /// @see pubnub_generate_uuid_v4_random
-    int set_uuid_v4_random()
+    PUBNUB_DEPRECATED int set_uuid_v4_random() 
+    {
+	return set_user_id_with_random_uuid_v4();
+    }
+
+    /// Set the user_id with a random-generated UUID
+    /// @see pubnub_generate_uuid_v4_random
+    int set_user_id_with_random_uuid_v4()
```
[the remark this answers is not in the document]
Comment (Xavrax) [reply cpp/pubnub_common.hpp]: I removed this functions - in addition I will remove them from examples to satisfy warrnings.

**Comment by budgetpreneur** [comment]
@Xavrax Where are we enforcing the input of _user_id_ or _uuid_?
When I run the bwlow sample code snippet, SDK runs without any _uuid_ query param in the request.
```
pubnub_init(pbp, "demo", "demo");
pubnub_publish(pbp, chan, "\"Hello world from sync!\"");
```

**Comment by budgetpreneur** [comment]
> @Xavrax Where are we enforcing the input of _user_id_ or _uuid_? When I run the below sample code snippet, SDK runs without any _uuid_ query param in the request.
> 
> ```
> pubnub_init(pbp, "demo", "demo");
> pubnub_publish(pbp, chan, "\"Hello world from sync!\"");
> ```

Looks like this issue is fixed. Now I see "Pubnub assert failed ..." message.

**Comment by budgetpreneur** [comment]
@Xavrax , I tried to re-run the Travis CI tests, but it again failed.

**Comment by Xavrax** [comment]
>  I tried to re-run the Travis CI tests, but it again failed.

@budgetpreneur I know what is wrong. Gonna fix it tonight.

**Comment by budgetpreneur** [comment]
@Xavrax  the AddressSanitizer issue I reported before, looks like still unaddressed.

**Comment by Xavrax** [comment]
@budgetpreneur Are we sure that we want to fix it here? This PR grew much bigger than I expected when I wanted to do this.
I would like to create another PR to address this trouble. It will be much easier to understand what happen.

**Review by budgetpreneur — APPROVED** [review]
LGTM. Done few manual tests. Travis CI tests passed.

**Comment by Xavrax** [comment]
Thank you!

**Comment by mateusz-dahlke** [comment]
@client-engineering-bot release as v4.0.0

**Pull request merged**